x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/SnapshotRetentionTask.java

```diff
@@ -247,21 +256,34 @@ void deleteSnapshots(Map<String, List<SnapshotInfo>> snapshotsToDelete,
 
         logger.info("starting snapshot retention deletion for [{}] snapshots", count);
         long startTime = nowNanoSupplier.getAsLong();
-        int deleted = 0;
+        final AtomicInteger deleted =  new AtomicInteger(0);
+        final AtomicInteger failed = new AtomicInteger(0);
         for (Map.Entry<String, List<SnapshotInfo>> entry : snapshotsToDelete.entrySet()) {
             String repo = entry.getKey();
             List<SnapshotInfo> snapshots = entry.getValue();
             for (SnapshotInfo info : snapshots) {
-                deleteSnapshot(getPolicyId(info), repo, info.snapshotId(), slmStats);
-                deleted++;
+                deleteSnapshot(getPolicyId(info), repo, info, slmStats, historyItem -> {
+                    // This would be nicer if we could use ifPresentOrElse
+                    historyItem.ifPresent(item -> {
+                        if (item.isSuccess()) {
+                            deleted.incrementAndGet();
+                        } else {
+                            failed.incrementAndGet();
+                        }
+                        historyStore.putAsync(item);
+                    });
+                    if (historyItem.isEmpty()) {
+                        failed.incrementAndGet();
+                    }
+                });
                 // Check whether we have exceeded the maximum time allowed to spend deleting
                 // snapshots, if we have, short-circuit the rest of the deletions
                 TimeValue elapsedDeletionTime = TimeValue.timeValueNanos(nowNanoSupplier.getAsLong() - startTime);
                 logger.trace("elapsed time for deletion of [{}] snapshot: {}", info.snapshotId(), elapsedDeletionTime);
                 if (elapsedDeletionTime.compareTo(maximumTime) > 0) {
                     logger.info("maximum snapshot retention deletion time reached, time spent: [{}]," +
-                            " maximum allowed time: [{}], deleted {} out of {} snapshots scheduled for deletion",
-                        elapsedDeletionTime, maximumTime, deleted, count);
+                            " maximum allowed time: [{}], deleted [{}] out of [{}] snapshots scheduled for deletion, failed to delete [{}]",
+                        elapsedDeletionTime, maximumTime, deleted, count, failed);
```

Comment:
I believe this needs to be `deleted.get()` and `failed.get()` now?

Reply:
This should be fine because `AtomicInteger#toString` returns the string representation of the value, but I'll double check how the message looks when you run it.